src/topology/continuous_function/basic.lean

```diff
@@ -224,6 +224,8 @@ def restrict (f : C(α, β)) : C(s, β) := ⟨f ∘ coe⟩
 
 @[simp] lemma coe_restrict (f : C(α, β)) : ⇑(f.restrict s) = f ∘ coe := rfl
 
+@[simp] lemma restrict_apply (f : C(α, β)) (s : set α) (x : s) : (f.restrict s) x = f x.val := rfl
```

Comment:
using `x.val` here is bad, as it is not the simp normal form. Use the coercion instead.

Reply:
The reason I wanted this is because several times in my code I need to do `rw` on expressions like `f.restrict S ⟨y, hy⟩`, where `(S : set α), (y : α), (hy : y ∈ α)`. I want to rewrite this into `f y` in a single step. It seems that @sgouezel's change wouldn't achieve this: it would instead rewrite this into `f ↑⟨y, hy⟩`, and `↑⟨y, hy⟩` is not syntatically equal to `y`, unlike `(⟨y, hy⟩).val`; so one needs to do a second rewrite with `subtype.coe_mk` in order to drill down to `f y`. 

The only reason to introduce this lemma was to make some rewrite chains shorter and easier to use: `rw continuous_map.restrict_apply` is a synonym for `rw [continuous_map.coe_restrict, function.comp_app, subtype.coe_mk]`. So making the chain longer again seems to defeat 50% of the purpose of introducing the lemma in the first place.

How about
```Lean
lemma restrict_apply_mk (f : C(α, β)) (s : set α) (x : α) (hx : x ∈ s) : 
  (f.restrict s) ⟨x, hx⟩ = f x.val := rfl
```
? That would be in simp-normal form, wouldn't it?

Reply:
Your suggestion has a typo, it should be `f x` on the RHS since `x` is no longer a subtype.

That would indeed be simp-normal form on the RHS, but if we did have it it should be in addition to the lemma that @sgouezel is requesting, not instead. (Your lemma is actually more useful for rewriting backwards, while the one being suggested is better for rewriting forwards)

> and `↑⟨y, hy⟩` is not syntactically equal to `y`, unlike `(⟨y, hy⟩).val`;

This is not true, neither is syntactically equal; the latter just happens to be equal under the reductions performed by `rw`.
In Lean 4 ``↑⟨y, hy⟩` and `(⟨y, hy⟩).val` are syntactically identical to each other though.

Reply:
> This is not true, neither is syntactically equal

OK, I'm no computer scientist, I just meant "equal enough to continue with `rw`" :-)

Reply:
For future reference, syntactically equal typically means "the underlying `expr`s (expression trees) are exactly the same by construction"

Reply:
OK, I stand corrected. So are we going to have either `restrict_apply` or `restrict_apply_mk`, or not? I'm more than happy to do without them in my code (particularly since it would avoid the rigmarole of forward-porting to mathlib4); I was just trying to be a good citizen by saving effort for future mathlib users / contributors.

Reply:
We should have either `restrict_apply` or both `restrict_apply` and `restrict_apply_mk`. You might find that `#lint` doesn't allow both to be tagged `simp`.

Reply:
Incidentally, I checked the library for comparable lemmas (to see if there was anything like `restrict_apply_mk` for other types). It seems not, which is probably a reasonable argument for not having it here either. But I did notice that
-  [subtype.restrict_apply](https://leanprover-community.github.io/mathlib_docs/data/subtype.html#subtype.restrict_apply) has `x.val` on the RHS, like my lemma above, so is presumably not in simp-NF either;
- [bounded_continuous_function.restrict_apply](https://leanprover-community.github.io/mathlib_docs/topology/continuous_function/bounded.html#bounded_continuous_function.restrict_apply) is something else entirely and should probably be renamed `coe_restrict`.

Reply:
The former looks like a mistake to me, while the latter is a missing feature  in `simps` which causes it to generate bad lemma names for unapplied function.